switches/http/http.cpp

```diff
@@ -0,0 +1,1280 @@
+    - GET /nodes, GET /nodes/:NODENAME with introspection
+    - POST /nodes/:NODENAME/:VARIABLE (sloppily allows GET /nodes/:NODENAME/:VARIABLE/:VALUE[/:VALUE]*)
+    - POST /nodes/:NODENAME/:EVENT (sloppily allows GET /nodes/:NODENAME/:EVENT[/:VALUE]*)
+    - form processing for updates and events (POST /.../:VARIABLE) and (POST /.../:EVENT)
+    - handle asynchronous variable reporting (GET /nodes/:NODENAME/:VARIABLE)
+    - JSON format for variable reporting (GET /nodes/:NODENAME/:VARIABLE)
+    - implement SSE streams and event filtering (GET /events) and (GET /nodes/:NODENAME/events)
+    - Aesl program bytecode upload (PUT /nodes/:NODENAME)
+      use curl --data-ascii "file=$(cat vmcode.aesl)" -X PUT http://127.0.0.1:3000/nodes/thymio-II
+    - accept JSON payload rather than HTML form for updates and events (POST /.../:VARIABLE) and (POST /.../:EVENT)
+ 
+ TODO:
+    - handle more than just one Thymio-II node
+    - gracefully shut down TCP/IP connections (half-close, wait, close)
+ 
+ This code borrows from the rest of Aseba, especially switches/medulla and examples/clients/cpp-shell,
+ which bear the copyright and LGPL 3 licensing notice below.
+ 
+ */
+/*
+	Aseba - an event-based framework for distributed robot control
+	Copyright (C) 2007--2012:
+ Stephane Magnenat <stephane at magnenat dot net>
+ (http://stephane.magnenat.net)
+ and other contributors, see authors.txt for details
+	
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Lesser General Public License as published
+	by the Free Software Foundation, version 3 of the License.
+	
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Lesser General Public License for more details.
+	
+	You should have received a copy of the GNU Lesser General Public License
+	along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <cstdlib>
+#include <cstring>
+#include <iostream>
+#include <sstream>
+#include <set>
+#include <valarray>
+#include <vector>
+#include <iterator>
+#include "http.h"
+#include "../../common/consts.h"
+#include "../../common/types.h"
+#include "../../common/utils/utils.h"
+#include "../../transport/dashel_plugins/dashel-plugins.h"
+
+#if defined(_WIN32) && defined(__MINGW32__)
+/* This is a workaround for MinGW32, see libxml/xmlexports.h */
+#define IN_LIBXML
+#endif
+#include <libxml/parser.h>
+#include <libxml/tree.h>
+
+#if DASHEL_VERSION_INT < 10003
+#	error "You need at least Dashel version 1.0.3 to compile Http"
+#endif // DAHSEL_VERSION_INT
+
+//== Some utility functions and diverse hacks ==============================================
+
+// because I don't understand locales and wide strings yet
+char* wstringtocstr(std::wstring w, char* name)
```

Comment:
Functions for string conversions are provided in `common/utils/utils.h`, is there a reason not to use them?